2003-11-18  Derek Price  <derek@ximbiot.com>

	* modules.c (do_module): Reject absolute paths.
	(Report and suggested fix from Tony Hoyle <tmh@nodomain.org>.)

Index: src/modules.c
===================================================================
RCS file: /cvs/ccvs/src/modules.c,v
retrieving revision 1.63.6.1
retrieving revision 1.63.6.2
diff -u -r1.63.6.1 -r1.63.6.2
--- src/modules.c	28 Apr 2003 18:52:15 -0000	1.63.6.1
+++ src/modules.c	18 Nov 2003 19:03:53 -0000	1.63.6.2
@@ -157,6 +157,17 @@
     }
 #endif
 
+    /* Don't process absolute directories.  Anything else could be a security
+     * problem.  Before this check was put in place:
+     *
+     *   $ cvs -d:fork:/cvsroot co /foo
+     *   cvs server: warning: cannot make directory CVS in /: Permission denied
+     *   cvs [server aborted]: cannot make directory /foo: Permission denied
+     *   $
+     */
+    if (isabsolute (mname))
+	error (1, 0, "Absolute module reference invalid: `%s'", mname);
+
     /* if this is a directory to ignore, add it to that list */
     if (mname[0] == '!' && mname[1] != '\0')
     {
